{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to be a Bioinformagician part 1\n",
    "- Author: Olga Botvinnik\n",
    "- Date: 2018-10-09\n",
    "\n",
    "Abstract: Many computational problems have already been solved and yet hundreds of hours are lost to re-solving them. This series provides tips and tricks that solve common pain points in bioinformatics, using AWS, reading/writing CSVs, extracting data out of file names, and lots more!\n",
    "\n",
    "## Prerequisities\n",
    "1. Everything you do should be full screen. Use [divvy](http://mizage.com/divvy/) to make managing windows easier - ask IT for a license\n",
    "1. [anaconda python](https://www.anaconda.com/download/#macos) - Install the 3.x (e.g. 3.7) version if you don't have it already\n",
    "1. aws account - Ask Olga or James for one\n",
    "1. awscli and aegea - on the command line, `pip install awscli aegea` after you install Anaconda Python\n",
    "1. nbconda - on the command line, `conda install nb_conda` after you install Anaconda Python\n",
    "1. reflow. Click this link: https://github.com/grailbio/reflow/releases/download/reflow0.6.8/reflow0.6.8.darwin.amd64 Then on the command line:\n",
    "    ```\n",
    "    cd ~/Downloads\n",
    "    chmod ugo+x reflow0.6.8.darwin.amd64\n",
    "    sudo mv reflow0.6.8.darwin.amd64 /usr/local/bin/reflow\n",
    "    ```\n",
    "    Now the command `reflow` should output a lot of stuff:\n",
    "    \n",
    "    ```\n",
    "     î‚° reflow\n",
    "    The reflow command helps users run Reflow programs, inspect their\n",
    "    outputs, and query their statuses.\n",
    "\n",
    "    The command comprises a set of subcommands; the list of supported\n",
    "    commands can be obtained by running\n",
    "\n",
    "        reflow -help\n",
    "\n",
    "    ... (more stuff) ...\n",
    "    ```\n",
    "    Then configure reflow, following the [Confluence entry on Reflow (https://czbiohub.atlassian.net/wiki/spaces/DS/pages/838205454/reflow) instructions for configuration:\n",
    "    ```\n",
    "    AWS_SDK_LOAD_CONFIG=1 reflow setup-ec2\n",
    "    AWS_SDK_LOAD_CONFIG=1 reflow setup-s3-repository czbiohub-reflow-quickstart-cache\n",
    "    AWS_SDK_LOAD_CONFIG=1 reflow setup-dynamodb-assoc czbiohub-reflow-quickstart\n",
    "    export AWS_REGION=us-west-2\n",
    "    ```\n",
    "1. Claim a folder within `s3://czbiohub-cupcakes/` with today's date and your username, e.g.:\n",
    "    ```\n",
    "    s3://czbiohub-cupcakes/2018-10-09/olgabot/\n",
    "    ```\n",
    "1. [GitHub](http://github.com) username and membership to [@czbiohub](https://github.com/czbiohub/) GitHub group.\n",
    "\n",
    "Highly recommended:\n",
    "- If you haven't seen it already, follow https://github.com/czbiohub/codonboarding\n",
    "- Especially \n",
    "    - install homebrew - it makes your life better for installing packages on mac\n",
    "    ```\n",
    "    /usr/bin/ruby -e \"$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)\"\n",
    "    ```\n",
    "    - install Oh My ZSH: https://ohmyz.sh/\n",
    "    - Install exa: https://the.exa.website/ - MUCH easier to install if you installed homebrew.\n",
    "    ```\n",
    "    brew install exa\n",
    "    ```\n",
    "\n",
    "\n",
    "## Installation\n",
    "\n",
    "### 1. Clone the cupcakes repo to `~/code` if you haven't already\n",
    "\n",
    "\n",
    "```\n",
    "mkdir ~/code\n",
    "cd ~/code\n",
    "git clone https://github.com/czbiohub/cupcakes\n",
    "```\n",
    "\n",
    "If you've already cloned it, change to the `master` branch and get the latest code with `git pull`:\n",
    "\n",
    "```\n",
    "cd ~/code/cupcakes/\n",
    "git checkout master\n",
    "git pull origin master\n",
    "```\n",
    "\n",
    "\n",
    "### 2. Create a \"bioinformagician\" environment\n",
    "\n",
    "```\n",
    "conda env create --name bioinformagician --file ~/code/cupcakes/2018/olgas_bioinformagician_tricks/environment.yml\n",
    "```\n",
    "\n",
    "You don't need to activate the environment because we are using `nbconda` to automatically add conda environments with Jupyter installed to the kernel list.\n",
    "\n",
    "## Running the notebook\n",
    "\n",
    "\n",
    "`cd` to your `~/code` directory in the terminal, then type `jupyter notebook`:\n",
    "\n",
    "```\n",
    "cd ~/code\n",
    "jupyter notebook\n",
    "```\n",
    "\n",
    "In the file browser, navigate to `cupcakes/2018/olgas_bioinformagician_tricks` and open `001_how_to_be_a_bioinformagician_part01.ipynb`\n",
    "\n",
    "\n",
    "In the notebook, make sure it is using the kernel named \"Python [conda env: bioinformagician]\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now we are ready to read the data!\n",
    "\n",
    "Run each cell below by pressing Shift+Enter\n",
    "\n",
    "\n",
    "1. Helpful jupter notebook keystrokes: \n",
    "    - The `Ctrl+M` means to press the Control and M letters together, then the space means to *let go* of those characters and press then next one(s).\n",
    "    - `Ctrl+M A` add cell above\n",
    "    - `Ctrl+M B` add cell below\n",
    "    - `Ctrl+M d d` delete cell\n",
    "    - `Ctrl+M i` interrupt\n",
    "    - `Ctrl+M 0` restart\n",
    "\n",
    "The ones above are the shortcuts I use the most. Go to Help > Keyboard Shortcuts to see them all."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standard convention is to import python standard libraries first, then third-party libraries after that\n",
    "# See list of standard libraries here: https://docs.python.org/3/library/\n",
    "# Both import lists should be alphabetically sorted\n",
    "\n",
    "# --- Python standard library --- #\n",
    "# Easily grab filenames from a folder\n",
    "import glob\n",
    "\n",
    "# Amazing library that I use almost every day for:\n",
    "# - chaining lists together into mega-lists\n",
    "# - \"multiplying\" lists against each other to get the full product of combinations\n",
    "import itertools\n",
    "\n",
    "# Read/write javascript object notation (JSON) files\n",
    "import json\n",
    "\n",
    "# Perform path manipulations\n",
    "import os\n",
    "\n",
    "# --- Third-party (non-standard Python) libraries --- #\n",
    "# python dataframes. very similar to R dataframes\n",
    "import pandas as pd\n",
    "\n",
    "# Make the number of characters allowed per column super big since our filenames are long\n",
    "pd.options.display.max_colwidth = 500"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read csv of lung cancer fastqs for which a kmer signature was calculated\n",
    "\n",
    "We previously computed kmer signatures for 5,000 lung cells, and the sample list is here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5054, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>read1</th>\n",
       "      <th>read2</th>\n",
       "      <th>name</th>\n",
       "      <th>output</th>\n",
       "      <th>trim_low_abundance_kmers</th>\n",
       "      <th>dna</th>\n",
       "      <th>protein</th>\n",
       "      <th>ksizes</th>\n",
       "      <th>scaled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B000419_S34/A10_B000419_S34_R1_001.fastq.gz</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B000419_S34/A10_B000419_S34_R2_001.fastq.gz</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>21,27,33,51</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A10_B000420_S82</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B000420_S82/A10_B000420_S82_R1_001.fastq.gz</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B000420_S82/A10_B000420_S82_R2_001.fastq.gz</td>\n",
       "      <td>A10_B000420_S82</td>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>21,27,33,51</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A10_B002073_S166</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002073_S166/A10_B002073_S166_R1_001.fastq.gz</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002073_S166/A10_B002073_S166_R2_001.fastq.gz</td>\n",
       "      <td>A10_B002073_S166</td>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>21,27,33,51</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A10_B002078_S202</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002078_S202/A10_B002078_S202_R1_001.fastq.gz</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002078_S202/A10_B002078_S202_R2_001.fastq.gz</td>\n",
       "      <td>A10_B002078_S202</td>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>21,27,33,51</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A10_B002095_S118</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002095_S118/A10_B002095_S118_R1_001.fastq.gz</td>\n",
       "      <td>s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002095_S118/A10_B002095_S118_R2_001.fastq.gz</td>\n",
       "      <td>A10_B002095_S118</td>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002095_S118.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>21,27,33,51</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id  \\\n",
       "0   A10_B000419_S34   \n",
       "1   A10_B000420_S82   \n",
       "2  A10_B002073_S166   \n",
       "3  A10_B002078_S202   \n",
       "4  A10_B002095_S118   \n",
       "\n",
       "                                                                                                                 read1  \\\n",
       "0    s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B000419_S34/A10_B000419_S34_R1_001.fastq.gz   \n",
       "1    s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B000420_S82/A10_B000420_S82_R1_001.fastq.gz   \n",
       "2  s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002073_S166/A10_B002073_S166_R1_001.fastq.gz   \n",
       "3  s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002078_S202/A10_B002078_S202_R1_001.fastq.gz   \n",
       "4  s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002095_S118/A10_B002095_S118_R1_001.fastq.gz   \n",
       "\n",
       "                                                                                                                 read2  \\\n",
       "0    s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B000419_S34/A10_B000419_S34_R2_001.fastq.gz   \n",
       "1    s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B000420_S82/A10_B000420_S82_R2_001.fastq.gz   \n",
       "2  s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002073_S166/A10_B002073_S166_R2_001.fastq.gz   \n",
       "3  s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002078_S202/A10_B002078_S202_R2_001.fastq.gz   \n",
       "4  s3://czbiohub-seqbot/fastqs/180516_A00111_0149_AH5CM2DSXX/rawdata/A10_B002095_S118/A10_B002095_S118_R2_001.fastq.gz   \n",
       "\n",
       "               name  \\\n",
       "0   A10_B000419_S34   \n",
       "1   A10_B000420_S82   \n",
       "2  A10_B002073_S166   \n",
       "3  A10_B002078_S202   \n",
       "4  A10_B002095_S118   \n",
       "\n",
       "                                                                 output  \\\n",
       "0   s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "1   s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "2  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "3  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature   \n",
       "4  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002095_S118.signature   \n",
       "\n",
       "   trim_low_abundance_kmers   dna  protein       ksizes  scaled  \n",
       "0                      True  True     True  21,27,33,51    1000  \n",
       "1                      True  True     True  21,27,33,51    1000  \n",
       "2                      True  True     True  21,27,33,51    1000  \n",
       "3                      True  True     True  21,27,33,51    1000  \n",
       "4                      True  True     True  21,27,33,51    1000  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_samples = pd.read_csv('lung_cancer/compute/samples.csv')\n",
    "print(compute_samples.shape)\n",
    "compute_samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read the documentation of the `sourmash_search.rf` file to see what we need\n",
    "\n",
    "We can use bash commands in Jupyter if we prefix them with a `!` (pronounced \"bang\") or excalamation point. Here, we're using the `reflow doc` command to show us the documentation of the program we want to run, `sourmash_search.rf`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parameters\r\n",
      "\r\n",
      "val signature string (required)\r\n",
      "    S3 path to single signature file e.g.\r\n",
      "    s3://olgabot-maca/facs/sourmash_compute_all/A1-B000610-3_56_F-1-1.sig\r\n",
      "val database string (required)\r\n",
      "    S3 full path to the sourmash database folder containing the database folder e.g.\r\n",
      "    s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/\r\n",
      "    Note: this folder contains tabula-muris-k21-protein.sbt.json and a bunch of\r\n",
      "    hidden files\r\n",
      "val database_name string (required)\r\n",
      "    Name of the database e.g.: tabula-muris-k21-protein\r\n",
      "val output string (required)\r\n",
      "    CSV file to write with search results e.g\r\n",
      "    s3://olgabot-maca/facs/sourmash_search/A1-B000610-3_56_F-1-1_tabula-muris-k21-protein.csv\r\n",
      "val ksize int = 21\r\n",
      "    Size of kmer to use (can only use one for index)\r\n",
      "val sequence_to_compare string = \"dna\"\r\n",
      "    What to compare, could be either \"protein\" or \"dna\"\r\n",
      "val ignore_abundance bool = false\r\n",
      "    Whether or not to include the abundance of kmers in the comparison\r\n",
      "\r\n",
      "Declarations\r\n",
      "\r\n",
      "val Search func(signature file, ksize int, database dir, database_name string) (csv file)\r\n",
      "\r\n",
      "val Main unit\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! reflow doc reflow/sourmash_search.rf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also look at the file `reflow/sourmash_search.rf` itself using `cat` (con***cat***enate the file to standard out) to see what's inside. Fortunately for us, `cat` is built-in to Jupyter so we can use it like we would on the command line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "param (\r\n",
      "    // S3 path to single signature file\r\n",
      "    // e.g. s3://olgabot-maca/facs/sourmash_compute_all/A1-B000610-3_56_F-1-1.sig\r\n",
      "    signature string\r\n",
      "\r\n",
      "    // S3 full path to the sourmash database folder containing the database folder\r\n",
      "    // e.g. s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/\r\n",
      "    // Note: this folder contains tabula-muris-k21-protein.sbt.json and a bunch of hidden files\r\n",
      "    database string\r\n",
      "\r\n",
      "    // Name of the database\r\n",
      "    // e.g.: tabula-muris-k21-protein\r\n",
      "    database_name string\r\n",
      "\r\n",
      "    // CSV file to write with search results\r\n",
      "    // e.g s3://olgabot-maca/facs/sourmash_search/A1-B000610-3_56_F-1-1_tabula-muris-k21-protein.csv\r\n",
      "    output string\r\n",
      "\r\n",
      "    // Size of kmer to use (can only use one for index)\r\n",
      "    ksize = 21\r\n",
      "\r\n",
      "    // What to compare, could be either \"protein\" or \"dna\"\r\n",
      "    sequence_to_compare = \"dna\"\r\n",
      "\r\n",
      "    // Whether or not to include the abundance of kmers in the comparison\r\n",
      "    ignore_abundance = false\r\n",
      ")\r\n",
      "\r\n",
      "\r\n",
      "val kmer_hashing = \"czbiohub/kmer-hashing:olgabot_search-compare-ignore-abundance\"\r\n",
      "\r\n",
      "// Instantiate the system modules \"files\" (system modules begin\r\n",
      "// with $), assigning its instance to the \"files\" identifier. To\r\n",
      "// view the documentation for this module, run \"reflow doc\r\n",
      "// $/files\".\r\n",
      "val files = make(\"$/files\")\r\n",
      "val dirs = make(\"$/dirs\")\r\n",
      "val path = make(\"$/path\")\r\n",
      "\r\n",
      "\r\n",
      "// Instantiate system module \"strings\"\r\n",
      "val strings = make(\"$/strings\")\r\n",
      "\r\n",
      "\r\n",
      "func Search(signature file, ksize int, database dir, database_name string) = {\r\n",
      "\r\n",
      "    val abundance_flag =  if ignore_abundance { \"--ignore-abundance\" } else { \"\" }\r\n",
      "\r\n",
      "\texec(image := kmer_hashing, mem := 24*GiB) (csv file) {\"\r\n",
      "        cd {{database}}\r\n",
      "        /opt/conda/bin/sourmash search \\\r\n",
      "            --threshold 0 \\\r\n",
      "            --output {{csv}} \\\r\n",
      "            {{abundance_flag}} \\\r\n",
      "            --ksize {{ksize}} \\\r\n",
      "            --{{sequence_to_compare}} \\\r\n",
      "            {{signature}} {{database_name}}\r\n",
      "    \"}\r\n",
      "}\r\n",
      "\r\n",
      "\r\n",
      "val sig = file(signature)\r\n",
      "val db = dir(database)\r\n",
      "\r\n",
      "val csv = Search(sig, ksize, db, database_name)\r\n",
      "\r\n",
      "val Main = files.Copy(csv, output)\r\n"
     ]
    }
   ],
   "source": [
    "cat reflow/sourmash_search.rf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will be creating a *Reflow batch job* which will submit 100s of lung cancer signatures to be looked up in the tabula muris database.\n",
    "\n",
    "\n",
    "For every argument, we'll need to create a column in a CSV that has that title, e.g. here we need the columns:\n",
    "\n",
    "- signature\n",
    "- database\n",
    "- database_name\n",
    "- output\n",
    "- ksize\n",
    "- sequence_to_compare\n",
    "- ignore_abundance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## look at aws folder of tabula muris kmer signature databases and save aws output to file\n",
    "\n",
    "This is where we start to do some magic. We can use *any* bash commands with `!`, including all `awscli` commands. Here, we `aws s3 ls` the `s3://olgabot-maca/facs/sourmash_index_all` folder, which we saved as the `prefix` variable, and access that variable in the \"bang\" statement with `$prefix`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                           PRE tabula-muris-k21-dna/\r\n",
      "                           PRE tabula-muris-k21-protein/\r\n",
      "                           PRE tabula-muris-k27-dna/\r\n",
      "                           PRE tabula-muris-k27-protein/\r\n",
      "                           PRE tabula-muris-k33-dna/\r\n",
      "                           PRE tabula-muris-k33-protein/\r\n",
      "                           PRE tabula-muris-k51-dna/\r\n",
      "                           PRE tabula-muris-k51-protein/\r\n"
     ]
    }
   ],
   "source": [
    "prefix = 's3://olgabot-maca/facs/sourmash_index_all'\n",
    "txt = 'sourmash_databases.txt'\n",
    "\n",
    "! aws s3 ls $prefix/ > $txt\n",
    "! cat $txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show that we now have a file called `sourmash_databases.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 5024\r\n",
      "drwxr-xr-x  14 olgabot  staff   476B Oct  9 16:04 \u001b[1m\u001b[36m.\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x  15 olgabot  staff   510B Oct  9 09:21 \u001b[1m\u001b[36m..\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x   4 olgabot  staff   136B Oct  8 14:12 \u001b[1m\u001b[36m.ipynb_checkpoints\u001b[m\u001b[m/\r\n",
      "-rw-r--r--   1 olgabot  staff   111K Oct  9 16:04 001_how_to_be_a_bioinformagician_part01.ipynb\r\n",
      "-rw-r--r--   1 olgabot  staff   2.3M Oct  9 09:44 002_how_to_be_a_bioinformagician_part02.ipynb\r\n",
      "drwxr-xr-x   4 olgabot  staff   136B Oct  8 18:24 \u001b[1m\u001b[36m__pycache__\u001b[m\u001b[m/\r\n",
      "-rw-r--r--   1 olgabot  staff   5.5K Oct  9 09:21 all_my_tricks_prep_notes.ipynb\r\n",
      "-rw-r--r--   1 olgabot  staff   2.9K Oct  9 09:21 environment.yml\r\n",
      "-rw-r--r--   1 olgabot  staff   1.3K Oct  8 18:20 extract_metadata.py\r\n",
      "drwxr-xr-x   3 olgabot  staff   102B Oct  9 09:44 \u001b[1m\u001b[36mlung-cancer\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x   3 olgabot  staff   102B Oct  9 09:42 \u001b[1m\u001b[36mlung_cancer\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x   3 olgabot  staff   102B Oct  9 09:21 \u001b[1m\u001b[36mreflow\u001b[m\u001b[m/\r\n",
      "-rw-r--r--   1 olgabot  staff   413B Oct  8 18:24 s3_utils.py\r\n",
      "-rw-r--r--   1 olgabot  staff   440B Oct  9 16:05 sourmash_databases.txt\r\n"
     ]
    }
   ],
   "source": [
    "ls -lha"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read the `aws s3 ls` output as a pandas table and do a bunch of transformations.\n",
    "\n",
    "We can extract the kmer size and the sequence of interest from the names of the database, plus add the full `s3` prefix before it so we can use it with our `sourmash_search.rf` script.\n",
    "\n",
    "Running each line one-by-one is left as an exercise to the reader :)\n",
    "\n",
    "- To uncomment each cell, put your cursor in it, then:\n",
    "    1. select-all with Command-A \n",
    "    2. uncomment with Command-/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# databases = pd.read_table(txt, delim_whitespace=True, header=None, names=['is_prefix', 'database_name'])\n",
    "# databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# databases['database_name'] = databases['database_name'].str.strip('/')\n",
    "# databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# databases = databases.drop('is_prefix', axis=1)\n",
    "# databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# databases['ksize'] = databases['database_name'].str.extract('k(\\d+)').astype(int)\n",
    "# databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# databases['sequence_to_compare'] = databases['database_name'].map(lambda x: x.split('-')[-1])\n",
    "# databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# databases['database'] = databases['database_name'].map(lambda x: f\"{prefix}/{x}/{x}/\")\n",
    "# databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# databases = databases.set_index('database_name')\n",
    "# databases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the table with pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ksize</th>\n",
       "      <th>sequence_to_compare</th>\n",
       "      <th>database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>database_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k21-dna</th>\n",
       "      <td>21</td>\n",
       "      <td>dna</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-dna/tabula-muris-k21-dna/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k21-protein</th>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k27-dna</th>\n",
       "      <td>27</td>\n",
       "      <td>dna</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-dna/tabula-muris-k27-dna/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k27-protein</th>\n",
       "      <td>27</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k33-dna</th>\n",
       "      <td>33</td>\n",
       "      <td>dna</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-dna/tabula-muris-k33-dna/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k33-protein</th>\n",
       "      <td>33</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k51-dna</th>\n",
       "      <td>51</td>\n",
       "      <td>dna</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-dna/tabula-muris-k51-dna/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k51-protein</th>\n",
       "      <td>51</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ksize sequence_to_compare  \\\n",
       "database_name                                         \n",
       "tabula-muris-k21-dna         21                 dna   \n",
       "tabula-muris-k21-protein     21             protein   \n",
       "tabula-muris-k27-dna         27                 dna   \n",
       "tabula-muris-k27-protein     27             protein   \n",
       "tabula-muris-k33-dna         33                 dna   \n",
       "tabula-muris-k33-protein     33             protein   \n",
       "tabula-muris-k51-dna         51                 dna   \n",
       "tabula-muris-k51-protein     51             protein   \n",
       "\n",
       "                                                                                                              database  \n",
       "database_name                                                                                                           \n",
       "tabula-muris-k21-dna              s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-dna/tabula-muris-k21-dna/  \n",
       "tabula-muris-k21-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/  \n",
       "tabula-muris-k27-dna              s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-dna/tabula-muris-k27-dna/  \n",
       "tabula-muris-k27-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/  \n",
       "tabula-muris-k33-dna              s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-dna/tabula-muris-k33-dna/  \n",
       "tabula-muris-k33-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/  \n",
       "tabula-muris-k51-dna              s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-dna/tabula-muris-k51-dna/  \n",
       "tabula-muris-k51-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "databases = pd.read_table(txt, delim_whitespace=True, header=None, names=['is_prefix', 'database_name'])\n",
    "databases['database_name'] = databases['database_name'].str.strip('/')\n",
    "databases = databases.drop('is_prefix', axis=1)\n",
    "databases['ksize'] = databases['database_name'].str.extract('k(\\d+)').astype(int)\n",
    "databases['sequence_to_compare'] = databases['database_name'].map(lambda x: x.split('-')[-1])\n",
    "databases['database'] = databases['database_name'].map(lambda x: f\"{prefix}/{x}/{x}/\")\n",
    "databases = databases.set_index('database_name')\n",
    "databases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Only compare on protein databases\n",
    "\n",
    "Since we're mapping human signatures onto a mouse database, we want to only compare on the protein signatures since protein sequences are more conserved than nucleotides. We use the [`query()` DataFrame method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.query.html) to get only the rows for which their `sequence_to_compare` is \"protein\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ksize</th>\n",
       "      <th>sequence_to_compare</th>\n",
       "      <th>database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>database_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k21-protein</th>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k27-protein</th>\n",
       "      <td>27</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k33-protein</th>\n",
       "      <td>33</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tabula-muris-k51-protein</th>\n",
       "      <td>51</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ksize sequence_to_compare  \\\n",
       "database_name                                         \n",
       "tabula-muris-k21-protein     21             protein   \n",
       "tabula-muris-k27-protein     27             protein   \n",
       "tabula-muris-k33-protein     33             protein   \n",
       "tabula-muris-k51-protein     51             protein   \n",
       "\n",
       "                                                                                                              database  \n",
       "database_name                                                                                                           \n",
       "tabula-muris-k21-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/  \n",
       "tabula-muris-k27-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/  \n",
       "tabula-muris-k33-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/  \n",
       "tabula-muris-k51-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_databases = databases.query('sequence_to_compare == \"protein\"')\n",
    "protein_databases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now comes the cool part!\n",
    "\n",
    "## \"Multiply\" the samples x databases x ignore abundances to get every combination\n",
    "\n",
    "We want to map each human lung sample onto ALL FOUR databases, PLUS we want to try both `ignore_abundance=True` and `ignore_abundance=False` \n",
    "\n",
    "Use `product` from Python's [itertools](https://docs.python.org/3/library/itertools.html) which is my favorite standard library module. (Though [collections](https://docs.python.org/3/library/collections.html) is a close second)\n",
    "\n",
    "Remember that Python was designed as \"batteries included\" so if you're doing something like doing a ton of nested for loops, know that many, many, *many* people have done that in the past and gotten confused/frustrated by it, too, and have figured out better ways to do it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature\n",
       "1     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature\n",
       "2    s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature\n",
       "3    s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature\n",
       "4    s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002095_S118.signature\n",
       "Name: output, dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_samples['output'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, False)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ignore_abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['tabula-muris-k21-protein', 'tabula-muris-k27-protein',\n",
       "       'tabula-muris-k33-protein', 'tabula-muris-k51-protein'],\n",
       "      dtype='object', name='database_name')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_databases.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(40432, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>signature</th>\n",
       "      <th>ignore_abundance</th>\n",
       "      <th>database_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40402</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40403</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40404</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40405</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40406</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40407</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40408</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40409</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40410</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40411</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40412</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40413</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40414</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40415</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40416</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40417</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40418</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40419</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40420</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40421</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40422</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40423</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40424</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40425</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40426</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40427</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40428</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40429</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40430</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40431</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40432 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  signature  \\\n",
       "0       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "1       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "2       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "3       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "4       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "5       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "6       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "7       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "8       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "9       s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "10      s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "11      s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "12      s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "13      s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "14      s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "15      s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000420_S82.signature   \n",
       "16     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "17     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "18     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "19     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "20     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "21     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "22     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "23     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002073_S166.signature   \n",
       "24     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature   \n",
       "25     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature   \n",
       "26     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature   \n",
       "27     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature   \n",
       "28     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature   \n",
       "29     s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B002078_S202.signature   \n",
       "...                                                                     ...   \n",
       "40402   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature   \n",
       "40403   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature   \n",
       "40404   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature   \n",
       "40405   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature   \n",
       "40406   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature   \n",
       "40407   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003570_S153.signature   \n",
       "40408   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature   \n",
       "40409   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature   \n",
       "40410   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature   \n",
       "40411   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature   \n",
       "40412   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature   \n",
       "40413   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature   \n",
       "40414   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature   \n",
       "40415   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003573_S237.signature   \n",
       "40416    s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature   \n",
       "40417    s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature   \n",
       "40418    s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature   \n",
       "40419    s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature   \n",
       "40420    s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature   \n",
       "40421    s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature   \n",
       "40422    s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature   \n",
       "40423    s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003586_S57.signature   \n",
       "40424   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature   \n",
       "40425   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature   \n",
       "40426   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature   \n",
       "40427   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature   \n",
       "40428   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature   \n",
       "40429   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature   \n",
       "40430   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature   \n",
       "40431   s3://olgabot-maca/lung_cancer/sourmash_v4/P9_B003588_S105.signature   \n",
       "\n",
       "       ignore_abundance             database_name  \n",
       "0                  True  tabula-muris-k21-protein  \n",
       "1                  True  tabula-muris-k27-protein  \n",
       "2                  True  tabula-muris-k33-protein  \n",
       "3                  True  tabula-muris-k51-protein  \n",
       "4                 False  tabula-muris-k21-protein  \n",
       "5                 False  tabula-muris-k27-protein  \n",
       "6                 False  tabula-muris-k33-protein  \n",
       "7                 False  tabula-muris-k51-protein  \n",
       "8                  True  tabula-muris-k21-protein  \n",
       "9                  True  tabula-muris-k27-protein  \n",
       "10                 True  tabula-muris-k33-protein  \n",
       "11                 True  tabula-muris-k51-protein  \n",
       "12                False  tabula-muris-k21-protein  \n",
       "13                False  tabula-muris-k27-protein  \n",
       "14                False  tabula-muris-k33-protein  \n",
       "15                False  tabula-muris-k51-protein  \n",
       "16                 True  tabula-muris-k21-protein  \n",
       "17                 True  tabula-muris-k27-protein  \n",
       "18                 True  tabula-muris-k33-protein  \n",
       "19                 True  tabula-muris-k51-protein  \n",
       "20                False  tabula-muris-k21-protein  \n",
       "21                False  tabula-muris-k27-protein  \n",
       "22                False  tabula-muris-k33-protein  \n",
       "23                False  tabula-muris-k51-protein  \n",
       "24                 True  tabula-muris-k21-protein  \n",
       "25                 True  tabula-muris-k27-protein  \n",
       "26                 True  tabula-muris-k33-protein  \n",
       "27                 True  tabula-muris-k51-protein  \n",
       "28                False  tabula-muris-k21-protein  \n",
       "29                False  tabula-muris-k27-protein  \n",
       "...                 ...                       ...  \n",
       "40402              True  tabula-muris-k33-protein  \n",
       "40403              True  tabula-muris-k51-protein  \n",
       "40404             False  tabula-muris-k21-protein  \n",
       "40405             False  tabula-muris-k27-protein  \n",
       "40406             False  tabula-muris-k33-protein  \n",
       "40407             False  tabula-muris-k51-protein  \n",
       "40408              True  tabula-muris-k21-protein  \n",
       "40409              True  tabula-muris-k27-protein  \n",
       "40410              True  tabula-muris-k33-protein  \n",
       "40411              True  tabula-muris-k51-protein  \n",
       "40412             False  tabula-muris-k21-protein  \n",
       "40413             False  tabula-muris-k27-protein  \n",
       "40414             False  tabula-muris-k33-protein  \n",
       "40415             False  tabula-muris-k51-protein  \n",
       "40416              True  tabula-muris-k21-protein  \n",
       "40417              True  tabula-muris-k27-protein  \n",
       "40418              True  tabula-muris-k33-protein  \n",
       "40419              True  tabula-muris-k51-protein  \n",
       "40420             False  tabula-muris-k21-protein  \n",
       "40421             False  tabula-muris-k27-protein  \n",
       "40422             False  tabula-muris-k33-protein  \n",
       "40423             False  tabula-muris-k51-protein  \n",
       "40424              True  tabula-muris-k21-protein  \n",
       "40425              True  tabula-muris-k27-protein  \n",
       "40426              True  tabula-muris-k33-protein  \n",
       "40427              True  tabula-muris-k51-protein  \n",
       "40428             False  tabula-muris-k21-protein  \n",
       "40429             False  tabula-muris-k27-protein  \n",
       "40430             False  tabula-muris-k33-protein  \n",
       "40431             False  tabula-muris-k51-protein  \n",
       "\n",
       "[40432 rows x 3 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ignore_abundances = True, False\n",
    "\n",
    "data = list(itertools.product(compute_samples['output'], ignore_abundances,\n",
    "                              protein_databases.index))\n",
    "\n",
    "samples = pd.DataFrame(data, columns=['signature', 'ignore_abundance', 'database_name'])\n",
    "print(samples.shape)\n",
    "samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract the `cell_id` from the signature filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>signature</th>\n",
       "      <th>ignore_abundance</th>\n",
       "      <th>database_name</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             signature  \\\n",
       "0  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "1  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "2  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "3  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "4  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "\n",
       "   ignore_abundance             database_name          cell_id  \n",
       "0              True  tabula-muris-k21-protein  A10_B000419_S34  \n",
       "1              True  tabula-muris-k27-protein  A10_B000419_S34  \n",
       "2              True  tabula-muris-k33-protein  A10_B000419_S34  \n",
       "3              True  tabula-muris-k51-protein  A10_B000419_S34  \n",
       "4             False  tabula-muris-k21-protein  A10_B000419_S34  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['cell_id'] = samples['signature'].map(lambda x: os.path.basename(x).split('.')[0])\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add an output location\n",
    "\n",
    "This is the full path to where we'll be storing the output csv files from the search results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv\n",
       "1    s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/A10_B000419_S34.csv\n",
       "2    s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/A10_B000419_S34.csv\n",
       "3    s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/A10_B000419_S34.csv\n",
       "4    s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv\n",
       "Name: output, dtype: object"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Change the following to the output location you chose! Should start with s3://czbiohub-cupcakes, \n",
    "# e.g. s3://czbiohub-cupcakes/2018-10-09/olgabot but use YOUR username! (not \"olgabot!\")\n",
    "output_prefix = 's3://czbiohub-cupcakes/2018-10-09/olgabot/'\n",
    "\n",
    "samples['output'] = output_prefix + samples['database_name'] + '/' + samples['cell_id'] + '.csv'\n",
    "samples['output'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add the protein database information to the samples\n",
    "We'll use the [`join()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.join.html) to combine the `protein_databases` table with our `samples` table, so we have the database `ksize`, `sequence_to_compare` and full URL.\n",
    "\n",
    "[Here](https://chrisalbon.com/python/data_wrangling/pandas_join_merge_dataframe/) is a nice blog post showing very clear examples of merge, join, and concatenate with pandas dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(40432, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>signature</th>\n",
       "      <th>ignore_abundance</th>\n",
       "      <th>database_name</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>output</th>\n",
       "      <th>ksize</th>\n",
       "      <th>sequence_to_compare</th>\n",
       "      <th>database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv</td>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/A10_B000419_S34.csv</td>\n",
       "      <td>27</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/A10_B000419_S34.csv</td>\n",
       "      <td>33</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/A10_B000419_S34.csv</td>\n",
       "      <td>51</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv</td>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             signature  \\\n",
       "0  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "1  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "2  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "3  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "4  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "\n",
       "   ignore_abundance             database_name          cell_id  \\\n",
       "0              True  tabula-muris-k21-protein  A10_B000419_S34   \n",
       "1              True  tabula-muris-k27-protein  A10_B000419_S34   \n",
       "2              True  tabula-muris-k33-protein  A10_B000419_S34   \n",
       "3              True  tabula-muris-k51-protein  A10_B000419_S34   \n",
       "4             False  tabula-muris-k21-protein  A10_B000419_S34   \n",
       "\n",
       "                                                                                   output  \\\n",
       "0  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv   \n",
       "1  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/A10_B000419_S34.csv   \n",
       "2  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/A10_B000419_S34.csv   \n",
       "3  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/A10_B000419_S34.csv   \n",
       "4  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv   \n",
       "\n",
       "   ksize sequence_to_compare  \\\n",
       "0     21             protein   \n",
       "1     27             protein   \n",
       "2     33             protein   \n",
       "3     51             protein   \n",
       "4     21             protein   \n",
       "\n",
       "                                                                                       database  \n",
       "0  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/  \n",
       "1  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/  \n",
       "2  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/  \n",
       "3  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/  \n",
       "4  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_databases = samples.join(protein_databases, on='database_name')\n",
    "print(samples_databases.shape)\n",
    "samples_databases.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a unique id for each row\n",
    "\n",
    "Reflow will create a log file for every job, and if you have duplicate ids, then those logs will get overwritten, and it won't treat those jobs as unique. So you want to have UNIQUE ids for each row.\n",
    "\n",
    "We use the [`apply()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.apply.html) to run the same function on every *row* of the dataframe. If we hadn't specified `axis=1`, it would have tried to apply the function to every *column* (`axis=0`). I usually forget which one is which so I try the function on one side and if it doesn't work I know it's the opposite."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>signature</th>\n",
       "      <th>ignore_abundance</th>\n",
       "      <th>database_name</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>output</th>\n",
       "      <th>ksize</th>\n",
       "      <th>sequence_to_compare</th>\n",
       "      <th>database</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv</td>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "      <td>A10_B000419_S34_ignore-abundance=True_tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/A10_B000419_S34.csv</td>\n",
       "      <td>27</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/</td>\n",
       "      <td>A10_B000419_S34_ignore-abundance=True_tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/A10_B000419_S34.csv</td>\n",
       "      <td>33</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/</td>\n",
       "      <td>A10_B000419_S34_ignore-abundance=True_tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/A10_B000419_S34.csv</td>\n",
       "      <td>51</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/</td>\n",
       "      <td>A10_B000419_S34_ignore-abundance=True_tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>A10_B000419_S34</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv</td>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "      <td>A10_B000419_S34_ignore-abundance=False_tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             signature  \\\n",
       "0  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "1  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "2  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "3  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "4  s3://olgabot-maca/lung_cancer/sourmash_v4/A10_B000419_S34.signature   \n",
       "\n",
       "   ignore_abundance             database_name          cell_id  \\\n",
       "0              True  tabula-muris-k21-protein  A10_B000419_S34   \n",
       "1              True  tabula-muris-k27-protein  A10_B000419_S34   \n",
       "2              True  tabula-muris-k33-protein  A10_B000419_S34   \n",
       "3              True  tabula-muris-k51-protein  A10_B000419_S34   \n",
       "4             False  tabula-muris-k21-protein  A10_B000419_S34   \n",
       "\n",
       "                                                                                   output  \\\n",
       "0  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv   \n",
       "1  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/A10_B000419_S34.csv   \n",
       "2  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/A10_B000419_S34.csv   \n",
       "3  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/A10_B000419_S34.csv   \n",
       "4  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/A10_B000419_S34.csv   \n",
       "\n",
       "   ksize sequence_to_compare  \\\n",
       "0     21             protein   \n",
       "1     27             protein   \n",
       "2     33             protein   \n",
       "3     51             protein   \n",
       "4     21             protein   \n",
       "\n",
       "                                                                                       database  \\\n",
       "0  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/   \n",
       "1  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/   \n",
       "2  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/   \n",
       "3  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/   \n",
       "4  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/   \n",
       "\n",
       "                                                                id  \n",
       "0   A10_B000419_S34_ignore-abundance=True_tabula-muris-k21-protein  \n",
       "1   A10_B000419_S34_ignore-abundance=True_tabula-muris-k27-protein  \n",
       "2   A10_B000419_S34_ignore-abundance=True_tabula-muris-k33-protein  \n",
       "3   A10_B000419_S34_ignore-abundance=True_tabula-muris-k51-protein  \n",
       "4  A10_B000419_S34_ignore-abundance=False_tabula-muris-k21-protein  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_databases['id'] = samples_databases.apply(lambda x: \n",
    "                              '{cell_id}_ignore-abundance={ignore_abundance}_{database_name}'.format(**x), \n",
    "                              axis=1)\n",
    "samples_databases.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subset to a few dozen cells\n",
    "We don't need to see the result of ALL human cells, which is ~5,000. We can just look at the output of a few to get a feel for how well it is working. Below are the sample ids that I chose, and we'll subset using the [`.query()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.query.html) and access the python variable with `@chosen_ids`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(168, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>signature</th>\n",
       "      <th>ignore_abundance</th>\n",
       "      <th>database_name</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>output</th>\n",
       "      <th>ksize</th>\n",
       "      <th>sequence_to_compare</th>\n",
       "      <th>database</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5568</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>C14_B003528_S62</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/C14_B003528_S62.csv</td>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "      <td>C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5569</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "      <td>C14_B003528_S62</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/C14_B003528_S62.csv</td>\n",
       "      <td>27</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/</td>\n",
       "      <td>C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5570</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "      <td>C14_B003528_S62</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/C14_B003528_S62.csv</td>\n",
       "      <td>33</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/</td>\n",
       "      <td>C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5571</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "      <td>C14_B003528_S62</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/C14_B003528_S62.csv</td>\n",
       "      <td>51</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/</td>\n",
       "      <td>C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5572</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>C14_B003528_S62</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/C14_B003528_S62.csv</td>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "      <td>C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                signature  \\\n",
       "5568  s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "5569  s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "5570  s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "5571  s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "5572  s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "\n",
       "      ignore_abundance             database_name          cell_id  \\\n",
       "5568              True  tabula-muris-k21-protein  C14_B003528_S62   \n",
       "5569              True  tabula-muris-k27-protein  C14_B003528_S62   \n",
       "5570              True  tabula-muris-k33-protein  C14_B003528_S62   \n",
       "5571              True  tabula-muris-k51-protein  C14_B003528_S62   \n",
       "5572             False  tabula-muris-k21-protein  C14_B003528_S62   \n",
       "\n",
       "                                                                                      output  \\\n",
       "5568  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/C14_B003528_S62.csv   \n",
       "5569  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/C14_B003528_S62.csv   \n",
       "5570  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/C14_B003528_S62.csv   \n",
       "5571  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/C14_B003528_S62.csv   \n",
       "5572  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/C14_B003528_S62.csv   \n",
       "\n",
       "      ksize sequence_to_compare  \\\n",
       "5568     21             protein   \n",
       "5569     27             protein   \n",
       "5570     33             protein   \n",
       "5571     51             protein   \n",
       "5572     21             protein   \n",
       "\n",
       "                                                                                          database  \\\n",
       "5568  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/   \n",
       "5569  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/   \n",
       "5570  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/   \n",
       "5571  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/   \n",
       "5572  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/   \n",
       "\n",
       "                                                                   id  \n",
       "5568   C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein  \n",
       "5569   C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein  \n",
       "5570   C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein  \n",
       "5571   C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein  \n",
       "5572  C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen_ids = ['C14_B003528_S62',\n",
    " 'D1_B003125_S25',\n",
    " 'E19_B003570_S199',\n",
    " 'F21_B000420_S213',\n",
    " 'G10_B003586_S142',\n",
    " 'G4_B003570_S232',\n",
    " 'G9_B003511_S57',\n",
    " 'H7_B003588_S211',\n",
    " 'I22_B002095_S22',\n",
    " 'I3_B003573_S63',\n",
    " 'J11_B003573_S95',\n",
    " 'J8_B003528_S224',\n",
    " 'K7_B002073_S103',\n",
    " 'L16_B003588_S16',\n",
    " 'L5_B003588_S5',\n",
    " 'M1_B000420_S61',\n",
    " 'M23_B002097_S251',\n",
    " 'N15_B000420_S99',\n",
    " 'O3_B003573_S207',\n",
    " 'P14_B000420_S146',\n",
    " 'P2_B003125_S14']\n",
    "\n",
    "samples_subset = samples_databases.query('cell_id in @chosen_ids')\n",
    "print(samples_subset.shape)\n",
    "samples_subset.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove sample ID column and set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(168, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>signature</th>\n",
       "      <th>ignore_abundance</th>\n",
       "      <th>database_name</th>\n",
       "      <th>output</th>\n",
       "      <th>ksize</th>\n",
       "      <th>sequence_to_compare</th>\n",
       "      <th>database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/C14_B003528_S62.csv</td>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k27-protein</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/C14_B003528_S62.csv</td>\n",
       "      <td>27</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k33-protein</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/C14_B003528_S62.csv</td>\n",
       "      <td>33</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>True</td>\n",
       "      <td>tabula-muris-k51-protein</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/C14_B003528_S62.csv</td>\n",
       "      <td>51</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein</th>\n",
       "      <td>s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature</td>\n",
       "      <td>False</td>\n",
       "      <td>tabula-muris-k21-protein</td>\n",
       "      <td>s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/C14_B003528_S62.csv</td>\n",
       "      <td>21</td>\n",
       "      <td>protein</td>\n",
       "      <td>s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                           signature  \\\n",
       "id                                                                                                                                     \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein   s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein   s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein   s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein   s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein  s3://olgabot-maca/lung_cancer/sourmash_v4/C14_B003528_S62.signature   \n",
       "\n",
       "                                                                 ignore_abundance  \\\n",
       "id                                                                                  \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein               True   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein               True   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein               True   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein               True   \n",
       "C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein             False   \n",
       "\n",
       "                                                                            database_name  \\\n",
       "id                                                                                          \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein   tabula-muris-k21-protein   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein   tabula-muris-k27-protein   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein   tabula-muris-k33-protein   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein   tabula-muris-k51-protein   \n",
       "C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein  tabula-muris-k21-protein   \n",
       "\n",
       "                                                                                                                                                 output  \\\n",
       "id                                                                                                                                                        \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein   s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/C14_B003528_S62.csv   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein   s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k27-protein/C14_B003528_S62.csv   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein   s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k33-protein/C14_B003528_S62.csv   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein   s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k51-protein/C14_B003528_S62.csv   \n",
       "C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein  s3://czbiohub-cupcakes/2018-10-09/olgabot/tabula-muris-k21-protein/C14_B003528_S62.csv   \n",
       "\n",
       "                                                                 ksize  \\\n",
       "id                                                                       \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein      21   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein      27   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein      33   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein      51   \n",
       "C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein     21   \n",
       "\n",
       "                                                                sequence_to_compare  \\\n",
       "id                                                                                    \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein              protein   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein              protein   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein              protein   \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein              protein   \n",
       "C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein             protein   \n",
       "\n",
       "                                                                                                                                                     database  \n",
       "id                                                                                                                                                             \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k21-protein   s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/  \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k27-protein   s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k27-protein/tabula-muris-k27-protein/  \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k33-protein   s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k33-protein/tabula-muris-k33-protein/  \n",
       "C14_B003528_S62_ignore-abundance=True_tabula-muris-k51-protein   s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k51-protein/tabula-muris-k51-protein/  \n",
       "C14_B003528_S62_ignore-abundance=False_tabula-muris-k21-protein  s3://olgabot-maca/facs/sourmash_index_all/tabula-muris-k21-protein/tabula-muris-k21-protein/  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_no_cell_id = samples_subset.drop(columns=['cell_id'])\n",
    "samples_no_cell_id = samples_no_cell_id.set_index('id')\n",
    "print(samples_no_cell_id.shape)\n",
    "samples_no_cell_id.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a folder to save reflow workflow to"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is our current directory?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/olgabot/code/cupcakes/2018/olgas_bioinformagician_tricks'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make the folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = 'lung_cancer/search_protein_databases'\n",
    "\n",
    "# The -p makes every parent folder recursively, in addition to your current folder\n",
    "! mkdir -p $folder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Double-check that the folder exists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 5032\r\n",
      "drwxr-xr-x  14 olgabot  staff   476B Oct  9 16:38 \u001b[1m\u001b[36m.\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x  15 olgabot  staff   510B Oct  9 09:21 \u001b[1m\u001b[36m..\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x   4 olgabot  staff   136B Oct  8 14:12 \u001b[1m\u001b[36m.ipynb_checkpoints\u001b[m\u001b[m/\r\n",
      "-rw-r--r--   1 olgabot  staff   116K Oct  9 16:38 001_how_to_be_a_bioinformagician_part01.ipynb\r\n",
      "-rw-r--r--   1 olgabot  staff   2.3M Oct  9 09:44 002_how_to_be_a_bioinformagician_part02.ipynb\r\n",
      "drwxr-xr-x   4 olgabot  staff   136B Oct  8 18:24 \u001b[1m\u001b[36m__pycache__\u001b[m\u001b[m/\r\n",
      "-rw-r--r--   1 olgabot  staff   5.5K Oct  9 09:21 all_my_tricks_prep_notes.ipynb\r\n",
      "-rw-r--r--   1 olgabot  staff   2.9K Oct  9 09:21 environment.yml\r\n",
      "-rw-r--r--   1 olgabot  staff   1.3K Oct  8 18:20 extract_metadata.py\r\n",
      "drwxr-xr-x   4 olgabot  staff   136B Oct  9 16:10 \u001b[1m\u001b[36mlung-cancer\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x   4 olgabot  staff   136B Oct  9 16:10 \u001b[1m\u001b[36mlung_cancer\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x   3 olgabot  staff   102B Oct  9 09:21 \u001b[1m\u001b[36mreflow\u001b[m\u001b[m/\r\n",
      "-rw-r--r--   1 olgabot  staff   413B Oct  8 18:24 s3_utils.py\r\n",
      "-rw-r--r--   1 olgabot  staff   440B Oct  9 16:05 sourmash_databases.txt\r\n"
     ]
    }
   ],
   "source": [
    "ls -lha"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write reflow batch `config.json` and `samples.csv` file\n",
    "\n",
    "We'll run the program `sourmash_search.rf` which is in the `reflow` folder here. I recommend keeping your reflow scripts separate from their batch folders as you may use the same script across multiple folders."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = \t{\n",
    "    # Since the folder we're writing to is relative to here as \"lung_cancer/search_protein_databases\"\n",
    "    # but the reflow folder is \"reflow/\" then we need to go up two directories with \"../..\"\n",
    "\t\"program\": \"../../reflow/sourmash_search.rf\",\n",
    "\t\"runs_file\": \"samples.csv\"\n",
    "\t}\n",
    "\n",
    "# Make sure the index (the ids!) are unique\n",
    "assert samples_no_cell_id.index.is_unique\n",
    "\n",
    "samples_no_cell_id.to_csv(f'{folder}/samples.csv', index=True)\n",
    "\n",
    "\n",
    "with open(f'{folder}/config.json', 'w') as f:\n",
    "    json.dump(config, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at the contents of the folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 128\r\n",
      "drwxr-xr-x  4 olgabot  staff   136B Oct  9 09:42 \u001b[1m\u001b[36m.\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x  4 olgabot  staff   136B Oct  9 16:10 \u001b[1m\u001b[36m..\u001b[m\u001b[m/\r\n",
      "-rw-r--r--  1 olgabot  staff    74B Oct  9 16:39 config.json\r\n",
      "-rw-r--r--  1 olgabot  staff    58K Oct  9 16:39 samples.csv\r\n"
     ]
    }
   ],
   "source": [
    "ls -lha $folder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Count the number of lines  in the folder to make sure it's the same as our input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     169 lung_cancer/search_protein_databases/samples.csv\r\n"
     ]
    }
   ],
   "source": [
    "! wc -l $folder/samples.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great, it's 168 rows + 1 header row = 169 rows!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## To save your changes for the future, create a branch and commit your changes\n",
    "\n",
    "Since you are saving the output to YOUR own bucket, you'll want to make sure you have the code that made these changes, and the best way to do that is to use `git`.\n",
    "\n",
    "Create a branch named like this: `yourgithubsername/bioinformagician-part1`, e.g.:\n",
    "\n",
    "```\n",
    "git checkout -b olgabot/bioinformagician-part1\n",
    "```\n",
    "\n",
    "Add all the files in the `olgas_bioinformagician_tricks` folder:\n",
    "\n",
    "```\n",
    "cd ~/code/cupcakes/2018/\n",
    "git add -A olgas_bioinformagician_tricks\n",
    "```\n",
    "\n",
    "Write a message about what files you're committing and why:\n",
    "\n",
    "```\n",
    "git commit -m \"Use a s3 bucket I can write to\"\n",
    "```\n",
    "\n",
    "Try to push the changes:\n",
    "\n",
    "```\n",
    "git push\n",
    "```\n",
    "\n",
    "Then you'll get a \"fatal error\" (but really nobody died so why the freakout?) that looks like this:\n",
    "\n",
    "\n",
    "```\n",
    "fatal: The current branch olgabot/enable_quality_filtering has no upstream branch.\n",
    "To push the current branch and set the remote as upstream, use\n",
    "\n",
    "    git push --set-upstream origin olgabot/enable_quality_filtering\n",
    "```\n",
    "\n",
    "\n",
    "Copy/paste THEIR `git push` command which will properly link up your own branch name with the remote branch name on GitHub. (This is what I always do... I'm too lazy to write out the full command myself)\n",
    "\n",
    "[Check out your branch in the whole tree here!](https://github.com/czbiohub/cupcakes/network)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running reflow jobs\n",
    "\n",
    "Finally, we're ready to run our 168 jobs! Reflow very nicely manages the jobs so we can \"set it and forget it\" while they run, and not have to worry about starting and stopping the instances.\n",
    "\n",
    "\n",
    "To run the reflow batch, go to the terminal and navigate to the `cupcakes/2018/olgas_bioinformagician_tricks/lung_cancer/search_protein_databases` directory. Once you're there, run this command:\n",
    "\n",
    "```\n",
    "reflow runbatch\n",
    "```\n",
    "\n",
    "You should see an output that looks like this:\n",
    "\n",
    "```\n",
    "  allocate {mem:24.0GiB cpu:1 disk:1.0GiB}[168]:  provisioning new instance                          1s\n",
    "  r4.xlarge:                                      awaiting fulfillment of spot request sir-18qg9fkp  1s\n",
    "  r4.xlarge:                                      awaiting fulfillment of spot request sir-fe6881cm  1s\n",
    "  r4.xlarge:                                      awaiting fulfillment of spot request sir-3q5g81vp  1s\n",
    "  r4.xlarge:                                      awaiting fulfillment of spot request sir-g51882ap  1s\n",
    "  r4.xlarge:                                      awaiting fulfillment of spot request sir-qff8azcn  1s\n",
    "batch /Users/olgabot/code/cupcakes/2018/olgas_bioinformagician_tricks/lung-cancer/search_protein_databases: remaining: 168\n",
    "  2b8d606f:  waiting  1s\n",
    "  ce90ab78:  waiting  1s\n",
    "  b301bf64:  waiting  1s\n",
    "  dbdb984c:  waiting  1s\n",
    "  c36afd43:  waiting  1s\n",
    "  b94b9ddd:  waiting  1s\n",
    "  3a866e4d:  waiting  1s\n",
    "  3b6ff8ab:  waiting  1s\n",
    "  52d5ecff:  waiting  1s\n",
    "  3838f6a3:  waiting  1s\n",
    "  9030d940:  waiting  1s\n",
    "  b352b3f3:  waiting  1s\n",
    "  a60c9a93:  waiting  1s\n",
    "  c9e1930f:  waiting  1s\n",
    "  6f47939f:  waiting  1s\n",
    "  f842c52f:  waiting  1s\n",
    "  a054367b:  waiting  1s\n",
    "```\n",
    "\n",
    "To stop the batch, hit Control+C to cancel it."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bioinformagician]",
   "language": "python",
   "name": "conda-env-bioinformagician-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
